configure.ac

```diff
@@ -61,4 +61,61 @@ AS_IF([test "x$enable_unit" != xno],
                          [AC_DEFINE([HAVE_CMOCKA],
                                     [1])])])
 AM_CONDITIONAL([UNIT], [test "x$enable_unit" != xno])
+
+AC_ARG_ENABLE([hardening],
+  [AS_HELP_STRING([--enable-hardening],
+    [Enable compiler and linker options to frustrate memory corruption exploits @<:@yes@:>@])],
+  [hardening="$enableval"],
+  [hardening="yes"])
+
+#
+# Good information on adding flags, and dealing with compilers can be found here:
+#   https://github.com/zcash/zcash/issues/1832
+#   https://github.com/kmcallister/autoharden/
+AS_IF([test x"$hardening" != x"no"], [
+
+  AC_DEFUN([add_hardened_c_flag], [
+    AX_CHECK_COMPILE_FLAG([$1],
+      [HARDENED_CFLAGS="$HARDENED_CFLAGS $1"],
+      [AC_MSG_ERROR([Cannot enable $1, consider configuring with --disable-hardening])]
+    )
+  ])
+
+  AC_DEFUN([add_hardened_ld_flag], [
+    AX_CHECK_LINK_FLAG([$1],
+      [HARDENED_LDFLAGS="$HARDENED_LDFLAGS $1"],
+      [AC_MSG_ERROR([Cannot enable $1, consider configuring with --disable-hardening])]
+    )
+  ])
+
+  add_hardened_c_flag([-Wall])
+  add_hardened_c_flag([-Wextra])
+  add_hardened_c_flag([-Werror])
+
+  add_hardened_c_flag([-Wformat])
+  add_hardened_c_flag([-Wformat-security])
+  add_hardened_c_flag([-Wstack-protector])
+  add_hardened_c_flag([-fstack-protector-all])
+
+  AX_CHECK_PREPROC_FLAG([-D_FORTIFY_SOURCE=2],[
+    AX_CHECK_PREPROC_FLAG([-U_FORTIFY_SOURCE],[
+      HARDENED_CPPFLAGS="$HARDENED_CPPFLAGS -U_FORTIFY_SOURCE"
+    ],[AC_MSG_ERROR(Cannot enable -U_FORTIFY_SOURCE)])
+    HARDENED_CPPFLAGS="$HARDENED_CPPFLAGS -D_FORTIFY_SOURCE=2"
+  ],[AC_MSG_ERROR(Cannot enable -D_FORTIFY_SOURCE=2)])
```

Comment:
I don't know if this nested logic buys us anything, we fail hard on each else case.. so likely just want to run them one at a time...